File: Backend/src/main/java/lan/dk/podcastserver/entity/Item.java
package lan.dk.podcastserver.entity;

import com.fasterxml.jackson.annotation.*;
import com.google.common.collect.Sets;
import javaslang.control.Option;
import javaslang.control.Try;
import lombok.*;
import lombok.experimental.Accessors;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.io.FilenameUtils;
import org.apache.commons.lang3.StringUtils;
import org.apache.commons.lang3.builder.HashCodeBuilder;
import org.hibernate.search.annotations.Boost;
import org.hibernate.search.annotations.DocumentId;
import org.hibernate.search.annotations.Field;
import org.hibernate.search.annotations.Indexed;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import javax.persistence.*;
import javax.validation.constraints.AssertTrue;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.ZonedDateTime;
import java.util.Set;
import java.util.UUID;

import static java.util.Objects.isNull;
import static java.util.Objects.nonNull;


@Entity
@Indexed
@Slf4j
@Builder
@Getter @Setter
@Table(name = "item", uniqueConstraints = @UniqueConstraint(columnNames={"podcast_id", "url"}))
@Accessors(chain = true)
@JsonIdentityInfo(generator = ObjectIdGenerators.PropertyGenerator.class, property = "id")
@NoArgsConstructor @AllArgsConstructor(access = AccessLevel.PRIVATE)
@JsonIgnoreProperties(ignoreUnknown = true, value = { "numberOfTry", "localUri", "addATry", "deleteDownloadedFile", "localPath", "proxyURLWithoutExtention", "extention", "hasValidURL", "reset", "coverPath" })
@EntityListeners(AuditingEntityListener.class)
public class Item {

    public  static Path rootFolder;
    public  static final Item DEFAULT_ITEM = new Item();
    private static final String PROXY_URL = "/api/podcasts/%s/items/%s/download%s";
    private static final String COVER_PROXY_URL = "/api/podcasts/%s/items/%s/cover.%s";

    @Id
    @DocumentId
    @GeneratedValue
    @Column(columnDefinition = "UUID")
    private UUID id;

    @OneToOne(fetch = FetchType.EAGER, cascade=CascadeType.ALL, orphanRemoval=true)
    private Cover cover;

    @ManyToOne(cascade={CascadeType.MERGE}, fetch = FetchType.EAGER)
    @JsonBackReference("podcast-item")
    private Podcast podcast;

    @Field @Boost(2.0F)
    @JsonView(ItemSearchListView.class)
    private String title;

    @Column(length = 65535)
    @JsonView(ItemSearchListView.class)
    private String url;

    @JsonView(ItemPodcastListView.class)
    private ZonedDateTime pubDate;

    @Field
    @Column(length = 2147483647)
    @JsonView(ItemPodcastListView.class)
    private String description;

    @JsonView(ItemSearchListView.class)
    private String mimeType;

    @JsonView(ItemDetailsView.class)
    private Long length;

    @JsonView(ItemDetailsView.class)
    private String fileName;

    /* Value for the Download */
    @Enumerated(EnumType.STRING)
    @JsonView(ItemSearchListView.class)
    private Status status = Status.NOT_DOWNLOADED;

    @Transient
    @JsonView(ItemDetailsView.class)
    private Integer progression = 0;

    @Transient
    private Integer numberOfTry = 0;

    @JsonView(ItemDetailsView.class)
    private ZonedDateTime downloadDate;

    @CreatedDate
    private ZonedDateTime creationDate;

    @JsonIgnore
    @ManyToMany(mappedBy = "items", cascade = CascadeType.REFRESH)
    private Set<WatchList> watchLists = Sets.newHashSet();


    public String getLocalUri() {
        return (fileName == null) ? null : getLocalPath().toString();
    }

    public Item setLocalUri(String localUri) {
        fileName = FilenameUtils.getName(localUri);
        return this;
    }

    public Item addATry() {
        this.numberOfTry++;
        return this;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Item)) return false;
        if (this == DEFAULT_ITEM && o != DEFAULT_ITEM || this != DEFAULT_ITEM && o == DEFAULT_ITEM) return false;

        Item item = Item.class.cast(o);

        if (nonNull(id) && nonNull(item.id))
            return id.equals(item.id);

        if (nonNull(url) && nonNull(item.url)) {
            return url.equals(item.url);
        }

        return StringUtils.equals(getProxyURL(), item.getProxyURL());
    }

    @Override
    public int hashCode() {
        return new HashCodeBuilder(17, 37)
                .append(url)
                .toHashCode();
    }

    @Override
    public String toString() {
        return "Item{" +
                "id=" + id +
                ", title='" + title + '\'' +
                ", url='" + url + '\'' +
                ", pubDate=" + pubDate +
                ", description='" + description + '\'' +
                ", mimeType='" + mimeType + '\'' +
                ", length=" + length +
                ", status='" + status + '\'' +
                ", progression=" + progression +
                ", downloaddate=" + downloadDate +
                ", podcast=" + podcast +
                ", numberOfTry=" + numberOfTry +
                '}';
    }

    @Transient @JsonProperty("proxyURL") @JsonView(ItemSearchListView.class)
    public String getProxyURL() {
        return String.format(PROXY_URL, podcast.getId(), id, getExtension());
    }

    @Transient @JsonProperty("isDownloaded") @JsonView(ItemSearchListView.class)
    public Boolean isDownloaded() {
        return StringUtils.isNotEmpty(fileName);
    }
    
    //* CallBack Method JPA *//
    @PreRemove
    public void preRemove() {
        checkAndDelete();
        watchLists.forEach(watchList -> watchList.remove(this));
    }

    private void checkAndDelete() {

        if (!podcast.getHasToBeDeleted())
            return;

        Try .of(() -> Files.deleteIfExists(getCoverPath()))
            .onFailure(e -> log.error("Error during deletion of cover of {}", this, e));

        if (isDownloaded()) {
            deleteFile();
        }
    }

    private void deleteFile() {
        Try.of(() -> Files.deleteIfExists(getLocalPath())).onFailure(e -> log.error("Error during deletion of {}", this, e));
    }

    @Transient @JsonIgnore
    public Item deleteDownloadedFile() {
        deleteFile();
        status = Status.DELETED;
        fileName = null;
        return this;
    }

    public Path getLocalPath() {
        return getPodcastPath().resolve(fileName);
    }

    public Path getCoverPath() {
        String url = isNull(cover) ? "" : cover.getUrl();
        return getPodcastPath().resolve(id + "." + FilenameUtils.getExtension(url));
    }

    private Path getPodcastPath() {
        return rootFolder.resolve(podcast.getTitle());
    }

    public String getProxyURLWithoutExtention() {
        return String.format(PROXY_URL, podcast.getId(), id, "");
    }

    private String getExtension() {
        return Option.of(fileName)
                .map(FilenameUtils::getExtension)
                .map(ext -> "."+ext)
                .getOrElse("");
    }

    @JsonProperty("cover") @JsonView(ItemSearchListView.class)
    public Cover getCoverOfItemOrPodcast() {
        return isNull(this.cover)
                ? podcast.getCover()
                : this.cover.toBuilder().url(String.format(COVER_PROXY_URL, podcast.getId(), id, FilenameUtils.getExtension(this.cover.getUrl()))).build();
    }

    @JsonProperty("podcastId") @JsonView(ItemSearchListView.class)
    public UUID getPodcastId() { return isNull(podcast) ? null : podcast.getId();}
        
    @AssertTrue
    public boolean hasValidURL() {
        return (!StringUtils.isEmpty(this.url)) || "send".equals(this.podcast.getType());
    }

    public Item reset() {
        checkAndDelete();
        setStatus(Status.NOT_DOWNLOADED);
        downloadDate = null;
        fileName = null;
        return this;
    }

    public interface ItemSearchListView {}
    public interface ItemPodcastListView extends ItemSearchListView {}
    public interface ItemDetailsView extends ItemPodcastListView {}
}

File: backend/src/main/kotlin/com/github/davinkevin/podcastserver/entity/Item.kt
package com.github.davinkevin.podcastserver.entity

import arrow.core.getOrElse
import arrow.core.toOption
import com.fasterxml.jackson.annotation.*
import com.github.davinkevin.podcastserver.utils.toVΛVΓ
import io.vavr.API.Option
import io.vavr.control.Option
import org.apache.commons.io.FilenameUtils
import org.apache.commons.lang3.StringUtils
import org.apache.commons.lang3.builder.HashCodeBuilder
import org.springframework.web.util.UriComponentsBuilder
import java.nio.file.Path
import java.time.ZonedDateTime
import java.util.*
import java.util.Objects.nonNull

@JsonIdentityInfo(generator = ObjectIdGenerators.PropertyGenerator::class, property = "id")
@JsonIgnoreProperties(ignoreUnknown = true, value = ["numberOfTry", "localUri", "addATry", "deleteDownloadedFile", "localPath", "proxyURLWithoutExtention", "extention", "hasValidURL", "reset", "coverPath"])
class Item {

    var id: UUID? = null
    var cover: Cover? = null

    @JsonBackReference("podcast-item")
    var podcast: Podcast? = null

    @JsonView(ItemSearchListView::class)
    var title: String? = null

    @JsonView(ItemSearchListView::class)
    var url: String? = null

    @JsonView(ItemPodcastListView::class)
    var pubDate: ZonedDateTime? = null

    @JsonView(ItemPodcastListView::class)
    var description: String? = null

    @JsonView(ItemSearchListView::class)
    var mimeType: String? = null

    @JsonView(ItemDetailsView::class)
    var length: Long? = null

    @JsonView(ItemDetailsView::class)
    var fileName: String? = null

    @JsonView(ItemSearchListView::class)
    var status = Status.NOT_DOWNLOADED

    @JsonView(ItemDetailsView::class)
    var progression: Int? = 0

    @JsonIgnore
    var numberOfFail: Int? = 0

    @JsonView(ItemDetailsView::class)
    var downloadDate: ZonedDateTime? = null

    var creationDate: ZonedDateTime? = null

    var localUri: String?
        get() = if (fileName == null) null else localPath.toString()
        set(value) { fileName = FilenameUtils.getName(value) }

    val proxyURL: String
        @JsonProperty("proxyURL")
        @JsonView(ItemSearchListView::class)
        get() = this.proxyURLWithoutExtention + extension

    val proxyURLWithoutExtention: String
        get() = UriComponentsBuilder.fromPath("/")
                .pathSegment(
                        "api",
                        "v1",
                        "podcasts",
                        podcast!!.id.toString(),
                        "items",
                        id.toString(),
                        Option<String>(title).map { s -> s.replace("[^a-zA-Z0-9.-]".toRegex(), "_") }.getOrElse("null")
                )
                .build(true)
                .toString()

    val isDownloaded: Boolean
        @JsonProperty("isDownloaded")
        @JsonView(ItemSearchListView::class)
        get() = StringUtils.isNotEmpty(fileName)

    val localPath: Path
        get() = podcastPath.resolve(fileName)

    val coverPath: Option<Path>
        get() = arrow.core.Option.fromNullable(cover)
                .flatMap { it.url.toOption() }
                .filter { !StringUtils.isEmpty(it)}
                .map { it.extension() }
                .map { podcastPath.resolve("${id.toString()}.$it") }
                .toVΛVΓ()
                .orElse { podcast!!.coverPath }

    private val podcastPath: Path
        get() = rootFolder!!.resolve(podcast!!.title)


    private val extension: String
        get() = Option<String>(fileName)
                .map { FilenameUtils.getExtension(it) }
                .map { ".$it" }
                .getOrElse("")

    val coverOfItemOrPodcast: Cover
        @JsonProperty("cover")
        @JsonView(ItemSearchListView::class)
        get() = Option<Cover>(cover)
                .map { c -> String.format(COVER_PROXY_URL, podcast!!.id, id, c.url!!.extension()) }
                .map { url ->
                    val c = Cover()
                    c.url = url
                    c.height = this.cover!!.height
                    c.width = this.cover!!.width
                    c.id = this.cover!!.id
                    c
                }
                .getOrElse { podcast!!.cover }

    val podcastId: UUID?
        @JsonProperty("podcastId")
        @JsonView(ItemSearchListView::class)
        get() = arrow.core.Option.fromNullable(podcast)
                .map { it.id }
                .getOrElse { null }

    fun addATry() {
        this.numberOfFail = (this.numberOfFail ?: 0) +1
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Item) return false
        if (this === DEFAULT_ITEM || other === DEFAULT_ITEM) return false

        if (nonNull(id) && nonNull(other.id))
            return id == other.id

        return if (nonNull(url) && nonNull(other.url)) url == other.url
        else StringUtils.equals(proxyURL, other.proxyURL)

    }

    override fun hashCode(): Int {
        return HashCodeBuilder(17, 37)
                .append(url)
                .toHashCode()
    }

    override fun toString(): String {
        return "Item{" +
                "id=" + id +
                ", title='" + title + '\''.toString() +
                ", url='" + url + '\''.toString() +
                ", pubDate=" + pubDate +
                ", description='" + description + '\''.toString() +
                ", mimeType='" + mimeType + '\''.toString() +
                ", length=" + length +
                ", status='" + status + '\''.toString() +
                ", progression=" + progression +
                ", downloaddate=" + downloadDate +
                ", podcast=" + podcast +
                ", numberOfTry=" + numberOfFail +
                '}'.toString()
    }

    interface ItemSearchListView
    interface ItemPodcastListView : ItemSearchListView
    interface ItemDetailsView : ItemPodcastListView

    companion object {
        private val log = org.slf4j.LoggerFactory.getLogger(Item::class.java)
        var rootFolder: Path? = null
        val DEFAULT_ITEM = Item().apply { id = UUID.randomUUID() }
        private const val COVER_PROXY_URL = "/api/v1/podcasts/%s/items/%s/cover.%s"
    }
}

private fun String.extension(): String {
    val extension = FilenameUtils.getExtension(this)
    return when {
        extension.isNullOrBlank() -> "jpg"
        else -> extension
    }
}
